tsdb/wal/checkpoint.go

```diff
@@ -238,6 +241,23 @@ func Checkpoint(logger log.Logger, w *WAL, from, to int, keep func(id chunks.Hea
 			}
 			stats.TotalExemplars += len(exemplars)
 			stats.DroppedExemplars += len(exemplars) - len(repl)
+		case record.Metadata:
+			metadata, err := dec.Metadata(rec, metadata)
+			if err != nil {
+				return nil, errors.Wrap(err, "decode metadata")
+			}
+			// Drop irrelevant metadata in place.
```

Comment:
Yeah, that's true. I thought this wouldn't have much impact as a) metadata isn't usually rapidly changing, and b) if the metadata _does_ change then it might make sense to be able to correlate it with the values associated with a series (eg. the unit changing from seconds to milliseconds).

If we _do_ want to keep the latest metadata entry only, we don't have any timestamps associated with them so I can't say how we'd go about it off the top of my head, but I'll take a look just in case we want to move to that direction.

Reply:
It makes sense to keep a metadata around until all the samples after that have been removed from the checkpoint until the new metadata appears for the reason you said - make sense of the old data. But I am not sure how efficiently we can do that. Once idea is to store the timestamp of the last sample seen when the metadata was updated.

> If we do want to keep the latest metadata entry only, we don't have any timestamps associated with them so I can't say how we'd go about it off the top of my head, but I'll take a look just in case we want to move to that direction.

This one is relatively simple but with one complexity. Similar to the `keep(ref)` method that we have, we have another `keepMeta(ref, meta)`, which returns true if the metadata matches the one present in the head block currently, i.e. the latest. Else false. Problem here is that, between 2 checkpoints, a series might have metadata M1, then switched to M2, then again back to M1, in which case we will end up keeping 2 copies of M1. We need to think how to dedupe this. Maybe a timestamp of some sort will help to identify the latest.

Reply:
The WAL is written to in-order, right? We can assume that for a given ref ID, metadata entries are written in-order and we can keep only the latest metadata entry for each ref ID. We shouldn't need to include timestamps, IIUC.

Reply:
Sorry for the late response. I think @rfratto's suggestion might work nicely.

One way to go about this would be to just keep track of the latest metadata for each series in the `for r.Next()` loop,  and then flush them separately along the remaining records in the end. I'll try it out to see how it works in practice, but I also wanted to hear how it sounds to you.

Reply:
I've played around with this approach and added a relevant test here https://github.com/tpaschalis/prometheus/commit/716d18bb1be1c5c2dafb6478302d5eef1cabc552. 

It seems to work nicely, will try to benchmark it and see how much of a difference it makes.

Reply:
If we do not care of old metadata, then definitely, just keep the last one!

Reply:
One of the use-cases brought up was keeping track of changing metadata. Do we not care about knowing what the series meant a month ago and only care about what it means now?